pkg/network/server.go

```diff
@@ -312,29 +315,37 @@ func (s *Server) handleMessage(peer Peer, msg *Message) error {
 		return errInvalidNetwork
 	}
 
-	switch msg.CommandType() {
-	case CMDAddr:
-		addrs := msg.Payload.(*payload.AddressList)
-		return s.handleAddrCmd(peer, addrs)
-	case CMDVersion:
-		version := msg.Payload.(*payload.Version)
-		return s.handleVersionCmd(peer, version)
-	case CMDHeaders:
-		headers := msg.Payload.(*payload.Headers)
-		go s.handleHeadersCmd(peer, headers)
-	case CMDInv:
-		inventory := msg.Payload.(*payload.Inventory)
-		return s.handleInvCmd(peer, inventory)
-	case CMDBlock:
-		block := msg.Payload.(*core.Block)
-		return s.handleBlockCmd(peer, block)
-	case CMDVerack:
-		// Make sure this peer has send his version before we start the
-		// protocol with that peer.
-		if peer.Version() == nil {
-			return errInvalidHandshake
+	if peer.Handshaked() {
+		switch msg.CommandType() {
+		case CMDAddr:
+			addrs := msg.Payload.(*payload.AddressList)
+			return s.handleAddrCmd(peer, addrs)
+		case CMDHeaders:
+			headers := msg.Payload.(*payload.Headers)
+			go s.handleHeadersCmd(peer, headers)
```

Comment:
curios: 
why do u handle headers in separate routine but handle blocks not?

Reply:
Well, it was done this way originally and I'm not yet sure why. Maybe it's worth some digging into, but I think not in this PR.

Reply:
I see